Create a documentation about PR (#5401)
vector-im/element-android: docs/pull_request.md

# Pull requests

## Introduction

This document gives some clue about how to efficiently manage Pull Requests (PR). This document is a first draft and may be improved later.

## Who should read this document?

Every pull request reviewers, but also probably every ones who submit PRs.

## Submitting PR

### Who can submit pull requests?

Basically every one who wants to contribute to the project! But there are some rules to follow.

#### Humans

> Developers who are employed (or paid) by Element can directly clone the project and push their branches and create PR.

Author: Will update, thanks

> Developers who are employed (or paid) by Element can directly clone the project and push their branches and create PR.

Author: done

External contributors must first fork the project and create PR to the mainline from there.

##### Draft PR?

Draft PR can be created when the submitter does not expect the PR to be reviewed and merged yet. It can be useful to publicly show the work, or to do a self-review first.

Draft PR can also be created when it depends on other un-merged PR.

In any case, it is better to explicitly declare in the description why the PR is a draft PR.

##### Assigned to reviewer?

> For the moment we do not directly assign PR to ask for a review. In the future we could have some automation to do that. Anyway the core team should have a look to every PR, even when not digging into details. More reviewers, more eys on the code change, more valuable feedback!

Reviewer: ```suggestion
For the moment we do not directly assign PR to ask for a review. In the future we could have some automation to do that. Anyway the core team should have a look to every PR, even when not digging into details. More reviewers, more eyes on the code change, more valuable feedback!
```

##### When create split PR?

To implement big new feature, it may be efficient to split the work into several smaller and scoped PRs. They will be easier to review, and they can be merged on develop faster.

Big PR can take time, and there is a risk of future merge conflict.

Feature flag can be used to avoid half implemented feature to be available in the application.

That said, splitting into several PRs should not have the side effect to have more review to do, for instance if some code is added, then finally removed.

##### Avoid unrelated fixing other issue in a big PR

> Each PR should focus on a single task. If other issues may be fixed when working in the area of it, it's preferable to open a dedicated PR.

Reviewer 2: A good example I think is file formatting. If formatting more than a line or two the file formatting can really obscure the intended purpose of code changes in a PR. Opening a PR that just formats a group of related files can make reviewers lives much easier.

Author: I agree; I will add it

Reviewer 3: formatting and noisy renames as separate PRs would be greatly appreciated :pray:

It will have the advantage to be reviewed and merged faster.

#### Bots

Some bots can create PR, but they still have to be reviewed by the team

##### Dependabot

Dependabot is a tool which maintain all our external dependencies up to date. A dedicated PR is created for each new available release for one of our external dependency.Dependabot

To review such PR, you have to
 - **IMPORTANT** check the diff files (as always).
 - Check the release note. Some existing bugs in Element project may be fixed by the upgrade
 - Make sure that the CI is happy
 - If the code does not compile (API break for instance), you have to checkout the branch and push new commits
 - Do some smoke test, depending of the library which has been upgraded

For some reason dependabot sometimes does not upgrade some dependencies. In this case, and when detected, the upgrade has to be done manually.

##### Gradle wrapper

> `Update Gradle Wrapper` is a tool which can create PR tu upgrade our gradle.properties file.

Reviewer 2: ```suggestion
`Update Gradle Wrapper` is a tool which can create PRs to upgrade our gradle.properties file.
```

Review such PR is the same recipe than for PR from Dependabot

##### Sync analytics plan

> This tools import any update in the analytics plan. See instruction in the PR itself to handle it.

Reviewer 2: ```suggestion
This tools imports any update in the analytics plan. See instruction in the PR itself to handle it.
```

Author: thanks

More info can be found in the file [analytics.md]

## Reviewing PR

### Who can review pull requests?

As an open source project, every one can review each others PR. Of course an approval from internal developer is mandatory for a PR to be merged.
But comment in PR from the community are always appreciated!

### What to have in mind when reviewing a PR

> 1. User experience: is the UX and UI correct? You will probably be the second person to test the new thing, the first one is the developer.

Reviewer 3: thinking out loud :smile: 

Ideally design/product would be involved/signed off before the change is implemented, letting us focus on the edgecases/potential errors in the implementation

It feels a little late to be questioning the core UX/UI after the change has been implemented :thinking: and potentially a little disheartening to find the thing you've built be rejected after investing time on it

2. Developer experience: does the code look nice and decoupled? No big functions, etc.
3. Code maintenance. A bit similar to point 2. Tricky code must be documented for instance

> 4. Fork consideration. Is configuration of forks will be easy? Some documentation may help in some cases.

Author: Note to myself: add a paragraph about the "Resolve conversation" thing. I do not like it because you have to open the conversation again to see the content, it's a waste of time.

Author: done

### Rules

#### Check the form

##### PR description

PR description should follow the PR template, and at least provide some context about the code change.

##### File change

1. Code should follow the guidelines
2. Code should be formatted correctly
3. XML attribute must be sorted
4. New code is added at the correct location
5. New classes are added to the correct location
6. Naming is correct. Naming is really important, it's considered part of the documentation
7. Architecture is followed. For instance, the logic is in the ViewModel and not in the Fragment

> 8. There is at least one file for the changelog. Exception if the PR fixes something which has not been released yet.

Reviewer 4: Maybe add a link here explaining how the `changelog` works, .misc .sdk .feature etc

Author: Yes I agree, as a general feedback, this file lacks lot of links to both internal and external locations.

> 9. PR includes test. allScreensTest when applicable, and unit tests

Reviewer: ```suggestion
9. PR includes tests. allScreensTest when applicable, and unit tests
```

😇

10. Avoid over complicating things. Keep it simple (KISS)!
11. PR contains only the expected change. Sometimes, the diff is showing changes that are already on develop. This is not good, submitter has to fix that up.

##### Check the commit

> Commit message must be short, one line and valuable. "WIP" is not a good commit message.

Reviewer 5: I wouldn't necessarily say they need to be short. Most git clients are okay with extra details being written on secondary lines.
Ideally they would also contain a link to the issue that it's referring to e.g. `element-android/issues/XXXX`

Also commit history should be nice. Having commits like "Adding temporary code" then later "Removing temporary code" is not good. The branch has to be rebased and those commit have to be dropped.

> PR merger could decide to squash and merge if commit history is not code.

Reviewer: ```suggestion
PR merger could decide to squash and merge if commit history is not good.
```
?

Commit like "Code review fixes" is good when reviewing the PR, since new changes can be reviewed easily, but is less valuable when looking at git history. I have no good solution to avoid this.

##### Check the substance

1. Test the changes!
2. Test the nominal case and the edge cases
3. Run the sanity test for critical PR

##### Make a dedicated meeting to review the PR

> Some times, big PR can be hard to review. Setting up a call with the PR submitter can faster the communication, rather than putting comment and questions as GitHub comment. It has the inconvenient to not make it public.

Reviewer 2: ```suggestion
Sometimes a big PR can be hard to review. Setting up a call with the PR submitter can speed up the communication, rather than putting comments and questions in GitHub comment. It has the inconvenience of making the discussion non-public, consider including a summary of the main points of the "offline" conversation in the PR.
```

### What happen to the issue(s)?

The issue(s) should be referenced in the PR description using keywords like `Closes` of `Fixes` followed by the issue number.

Example:
> Closes #1

Note that you have to repeat the keyword in case of a list of issue

> Closes #1, Closes #2, etc.

When PR will be merged, such referenced issue will be automatically closed.

> It up to the person who has merged the PR to go to the (closed) issue(s) and to add a comment to inform in which version the issue fix will be available. Use the current version of develop branch.

Reviewer: ```suggestion
It is up to the person who has merged the PR to go to the (closed) issue(s) and to add a comment to inform in which version the issue fix will be available. Use the current version of develop branch.
```

> > Closed in Element Android v1.x.y

Reviewer: FTR, we do not have that on that on other projects (EI, EW, Synapse).

Author: Yes, but I think it is very valuable for issue submitters and for every one to see in this issue timeline in which version of the app the issue is actually fixed. We could also decide to close issue only once the code is released to avoid misunderstanding, but it's breaking automation (issue closed when merging PR).
I would suggest that other projects should do that too :)

### Merge conflict

> It's up to the submitter to handle merge conflict. Sometimes, they can be fixed directly from GitHub, sometimes this is not possible. The branch can be rebased on develop, or develop can be merged on the branch, it's up to the submitter to decide what is best.

Reviewer: We could add that GH actions do not run if there are merge conflicts.

Author: will do, thanks.

### When and who can merge PR

> PR can be merged by the submitter, if and only if at least one approval from another developer is done. Approval from all people added as reviewer is also a good thing to have. Approval from design team may be mandatory, but is not sufficient to merge a PR.

Reviewer: Should it be "PR **must** be merged by the submitter" ? For submitters who can, those who are maintainer of the repo.
Sometimes submitters forget that they have to commit.

Author: PR should be merged by code owners (or project owners) and not submitters from my point of view. Code owners could decide what can be part of the next release or not, and judge the "risk" of merging a PR regarding the time left until the next scheduled release.

Reviewer 3: I wonder if it's worth introducing a label for _merge in next release cycle_ so that we could approve the PRs and provide more visibility for the anticipated merge window

# Pull requests

## Introduction

This document gives some clue about how to efficiently manage Pull Requests (PR). This document is a first draft and may be improved later.

## Who should read this document?

Every pull request reviewers, but also probably every ones who submit PRs.

## Submitting PR

### Who can submit pull requests?

Basically every one who wants to contribute to the project! But there are some rules to follow.

#### Humans

People with write access to the project can directly clone the project, push their branches and create PR.

External contributors must first fork the project and create PR to the mainline from there.

##### Draft PR?

Draft PR can be created when the submitter does not expect the PR to be reviewed and merged yet. It can be useful to publicly show the work, or to do a self-review first.

Draft PR can also be created when it depends on other un-merged PR.

In any case, it is better to explicitly declare in the description why the PR is a draft PR.

Also, draft PR should not stay indefinitely in this state. It may be removed if it is the case and the submitter does not update it after a few days.

> ##### PR Review Assignment

Reviewer: ```suggestion
## PR Review Assignment
```

I think this section deserves higher visibility. It has a huge impact on daily work. We should be able to send people directly to this section and the title must be more visible.

Author: I agree, but it impacts the whole toc. I can update that later

We use automatic assignment for PR reviews. A PR is automatically routed by GitHub to a team member using the round robin algorithm. The process is the following:

> - The PR creator assigns the [element-android](https://github.com/orgs/vector-im/teams/element-android) team as a reviewer. They can skip this process and assign directly a specific member if they think they should take a look at it.

Reviewer 3: thinking out loud (not a blocker for this PR) 

could we automate this assignment? similar to how we apply the triage project to new issues~
I'm happy to have a go at updating the github actions

Reviewer: This is a good idea but we still want to be able to assign someone directly. If we change the selected reviewer afterwards, it will create noise.

There is a missing paragraph from the original doc that was used to create this section. I will add a comment.

Reviewer 3: I wouldn't mind adding specific reviewers as additional reviewers :see_no_evil: I wonder if when a specific person is being added as a reviewer it's a sign that the area could benefit from more eyes for sharing the knowledge

> - GitHub automatically assigns one reviewer. If the chosen reviewer is not available (holiday, etc.), remove them and manually add another reviewer.

Reviewer: ```suggestion
- GitHub automatically assigns one reviewer. If the chosen reviewer is not available (holiday, etc.), remove them and set again the team, GitHub will select another reviewer.
```

Author: done

- The reviewer gets a notification to make the review: they review the code following the good practice (see the rest of this document).
- After making their own review, if they feel not confident enough, they can ask another person for a full review, or they can tag someone within a PR comment to check specific lines.

For PRs coming from the community, the issue wrangler can assign either the team [element-android](https://github.com/orgs/vector-im/teams/element-android) or any member directly.

> 

Reviewer: What about this paragraph from the original doc?
```
If a PR depends on another PR, assign the same reviewer, even if they are from different repositories. For example, related matrix-ios-sdk and element-ios PRs should be reviewed by the same person. Run the automatic assignment on the first PR and use the selected reviewer for the second PR.
```

##### Gradle wrapper

`Update Gradle Wrapper` is a tool which can create PR to upgrade our gradle.properties file.
Review such PR is the same recipe than for PR from Dependabot

##### Sync analytics plan

This tools imports any update in the analytics plan. See instruction in the PR itself to handle it.
More info can be found in the file [analytics.md]

## Reviewing PR

### Who can review pull requests?

As an open source project, every one can review each others PR. Of course an approval from internal developer is mandatory for a PR to be merged.
But comment in PR from the community are always appreciated!

### What to have in mind when reviewing a PR

1. User experience: is the UX and UI correct? You will probably be the second person to test the new thing, the first one is the developer.
2. Developer experience: does the code look nice and decoupled? No big functions, new classes added to the right module, etc.
3. Code maintenance. A bit similar to point 2. Tricky code must be documented for instance
4. Fork consideration. Will configuration of forks be easy? Some documentation may help in some cases.

> 5. We are building long term products. "Quick and dirty" code must be avoided.

Reviewer 3: do you think it's worth including some examples? I've noticed a trend in community PRs commenting out existing logic rather than deleting the no longer used code   

this could probably be a separate doc for coding guidelines :sweat_smile:

6. The PR includes new tests for the added code, updated test for the existing code
7. All PRs from external contributors **MUST** include a sign-off. It's in the checklist, and sometimes it's checked by the submitter, but there is actually no sign-off. In this case, ask nicely for a sign-off and request changes (do not approve the PR, even if everything else is fine).

### Rules

#### Check the form

> ##### PR description

Reviewer 3: do you think we should have some conventions for the PR title as well?

Author: will add thanks

PR description should follow the PR template, and at least provide some context about the code change.

##### File change

1. Code should follow the guidelines
2. Code should be formatted correctly

> 3. XML attribute must be sorted

Reviewer 3: will the CI - lint/code quality script catch these? we might be able to delegate these points to the guidelines docs  (would be handy if the guidelines was a link)

### What happen to the issue(s)?

The issue(s) should be referenced in the PR description using keywords like `Closes` of `Fixes` followed by the issue number.

Example:
> Closes #1

Note that you have to repeat the keyword in case of a list of issue

> Closes #1, Closes #2, etc.

When PR will be merged, such referenced issue will be automatically closed.
It is up to the person who has merged the PR to go to the (closed) issue(s) and to add a comment to inform in which version the issue fix will be available. Use the current version of `develop` branch.

> Closed in Element Android v1.x.y

### Merge conflict

It's up to the submitter to handle merge conflict. Sometimes, they can be fixed directly from GitHub, sometimes this is not possible. The branch can be rebased on `develop`, or the `develop` branch can be merged on the branch, it's up to the submitter to decide what is best.
Keep in mind that Github Actions are not run in case of conflict.

> ### When and who can merge PR

Reviewer 3: is it worth calling out how we merge? eg is there a preference for merging PRs as a squash/merge~ or the different scenarios where we might use one or the other (small fixes can be nice to squash)
